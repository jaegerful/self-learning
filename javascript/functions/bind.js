/* javascript does not use dynamic binding for normal variables. */

function outer() {
    let x = 3

    function useouter() {
        return 1 + x
    }

    function useinner() {
        let x = 0
        return useouter()
    }


    return useinner()
}

console.log(outer())

/* the main exceptions are 'this' and 'arguments'. */

/* 'this' cannot be changed within a function. */

const a = {name: 'a'}
const b = {name: 'b'}

/* instead of using 'func.call' and 'func.apply' */
/* for the same object over and over again, call */
/* 'func.bind' once. it returns a bound version */
/* of a function. */

const x = function() {
    console.log(this.name)
}

const y = x.bind(a)
y() /* all calls to 'y' have 'this' set to 'a'. */

/* 'func.bind' also accepts arguments. they are the */
/* first passed arguments to the function. */
const z = function(a, b) {
    console.log(a + b)
}

const c = z.bind(null, 2)
c(3) /* all calls to 'c' have first parameter set to 2. */

/* 'this' is permanently set by 'func.bind'. even */
/* if a method is called through its parent object, */
/* 'this' won't change. */
const d = {name: 'mr.'}

a.hi = function() {console.log(`hi, ${this.name}`)}
a.hi = a.hi.bind(d)

a.hi()