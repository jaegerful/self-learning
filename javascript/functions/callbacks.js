/* asynchronous code - code that does something concurrent to the main thread */
/* (i.e., concurrent to the script file). */

/* for example, scheduling a function to run later with 'setTimeout' */
/* is asynchronous. you're scheduling a function to run concurrently */
/* in the future. */

setTimeout(() => {
    console.log(`this is an asynchronous function!`)
}, 1000)

/* you may want to do something after asynchronous code executes. */
/* one approach is to provide callbacks for event listeners. */

function loadScript(src, callback) {
    const script = document.createElement('script');
    script.src = src;
  
    script.onload = () => callback(null, script);
    script.onerror = () => callback(new Error(`script load error for ${src}.`));
  
    document.head.append(script);
}

loadScript('./class-inheritance.js', function(error, script) {
    /* callback for 'loadScript'. */
    
    if (error) {
        console.log(`${error.name}: ${error.messsage}`)
        return
    }

    console.log(`script loaded successfully!`)
    /* possibly load another script... */
})

/* 'callback hell' or the 'pyramid of doom' occurs when */
/* callbacks have their own callbacks, and so on... */

/* this happens when we have a chain of asynchronous code, */
/* which need to execute in sequence, after each piece */
/* finishes executing. */

/* for this reason, the recursive approach */
/* to chain asynchronous code is frowned on. */