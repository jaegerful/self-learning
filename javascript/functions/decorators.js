/* decorators return a wrapper for a passed function. */
let inc = (a) => a + 1

/* the wrapper adds to the behavior of the function. */
/* in this case, integration with a cache. */
const cachingDecorator = (f) => {
    let cache = new Map()

    return function(a) {
        if (cache.has(a))
            return cache.get(a)

        let result = f(a)
        cache.set(a, result)
        return result
    }
}

/* does not override 'f' call. 'f' still exists in lexical environment. */
inc = cachingDecorator(inc)

console.time('inc(50): first call')
inc(50)
console.timeEnd('inc(50): first call')

console.time('inc(50): second call')
inc(50)
console.timeEnd('inc(50): second call')

console.log()

console.time('inc(1031): first call')
inc(1004)
console.timeEnd('inc(1031): first call')

console.time('inc(1031): second call')
inc(1004)
console.timeEnd('inc(1031): second call')

console.log()

/* even though 'this' is a part of lexical environments, it is not */
/* traced for if undefined. thus, every function needs its own 'this'. */

/* for this reason, 'function.call' exists. it accepts an arbitrary. */
/* number of arguments. */

let obj = {x: 3}

const y = function(a, b) {return a * b  * this.x}
let result = y.call(obj, ...[2, 4]) /* try 'y()' to get an error. */

console.log(`result of y.call: ${result}`)

/* 'function.apply' is identical. the only difference is that it */
/* accepts any array-like object for its arguments. */

result = y.apply(obj, [2, 4])
console.log(`result of y.call: ${result}`)

/* parameters are declared by name for easy access. */
/* however. there are other ways to access arguments. */

/* functions support the 'arguments' array-like object. */
/* here, arguments can be accessed by index. */

/* you can pass an arbitrary number of arguments */
/* to a function. only the needed ones will be used. */