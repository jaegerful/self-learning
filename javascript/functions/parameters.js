/* objects are passed by-reference. */
/* this includes functions and arrays. */