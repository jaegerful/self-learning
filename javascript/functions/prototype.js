/* a constructor's 'prototype' property is used as */
/* the prototype for newwly-created objects. */

const creature = {
    breathes: true,
}

function Rabbit(name) {
    if (new.target) /* if used as constructor. */
        this.name = name
}

Rabbit.prototype = creature

const Peter = new Rabbit('Peter Rabbit')
console.log(`${Peter.name} breathes: ${Peter.breathes}.`)

/* it's important to note that the constructor itself doesn't */
/* inherit from 'prototype'. only the new object does. */

console.log()

/* furthermore, every function has a default 'prototype' value. */
/* it's an object with one property: 'constructor'. */

/* 'constructor' is a reference to the function itself. */
/* here's a visualization: */

/* creature.prototype = {constructor: creature} */

/* in other words, newly-created objects would inherit */
/* from an object with one property: 'constructor'. */
/* they don't inherit from actual constructor, though. */

const Terminator = function() {
    this.fights = true
}

const descriptors = Object.getOwnPropertyDescriptors(Terminator.prototype)

console.log(`Terminator.prototype inherits from Object.prototype: ${Terminator.prototype.__proto__ === Object.prototype}.\n`)
console.log(`properties in default Terminator.prototype:`)

for (let key in descriptors)
    console.log(`${key}: ${Terminator.prototype[key]}`)

/* the 'constructor' property can be used instead of the actual constructor. */

let first = new Terminator()
first.fights = false

let second = new (first.constructor)() /* 'new' context supersedes context set by 'first'. */

console.log(`first.fights === false: ${first.fights === false}`)
console.log(`second.fights === true: ${second.fights === true}`)

console.log();

/* if called without 'new', 'first' is modified by constructor again. */
(first.constructor)()
console.log(`first.fights === true: ${first.fights === true}`)