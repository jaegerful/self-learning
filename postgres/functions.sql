-- $$ are the modern way to initialize strings.

-- parameter modes: 
-- 1. IN (default) - constants
-- 2. OUT - parameters that are all returned by function as a row. cannot be passed in call.
-- 3. INOUT - same as OUT parameters, but must also be assigned a value at function call.

CREATE OR REPLACE FUNCTION insert_entry(IN something INTEGER)
    RETURNS INTEGER -- can be ommited when using OUT or INOUT parameters
    LANGUAGE plpgsql
    AS
$$

    <<custom_name>> -- beginning a plpgsql block.
    DECLARE
        count INTEGER; -- simple declaration.
        id_variable public.person.id%type; -- assignment based on tables or other variables.
        row_variable public.person%rowtype; -- a variable that can hold an entire row from said table.
        another_row_variable record; -- 'record' is a variable for a row, without a predefined structure. it can be reassigned to any structure.
        a_constant CONSTANT INTEGER = 23; -- a constant variable that cannot change. 

    BEGIN -- not a transaction, but a plpgsql block.
        count = 2 -- variables can be altered in the body of the function :D... duh...
        SELECT * FROM public.person INTO id_variable WHERE id = 1;
        RETURN count;
    END custom_name; -- ends block.

$$; -- end of function definition (which could be in one line).

-- global variable 'found' set to true if previous SELECT INTO statement yielded at least one entry.