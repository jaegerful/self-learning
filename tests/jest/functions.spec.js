/* each file is a test suite. each test function is a test case (which can be composed of multiple expectations). */

import functions from './functions.js'

test('2 + 2 = 4', () => expect(functions.add(2, 2)).toBe(4)) // expect() returns an object.
test('2 + 2 != 4', () => expect(functions.add(2, 2)).not.toBe(5))
test('\'retnull\' returns null', () => expect(functions.retnull()).toBe(null))
test('\'reflect(false)\' is falsy', () => expect(functions.reflect(false)).toBeFalsy())
test('\'create()\' should have my information', () => expect(functions.create()).toEqual({'first': 'Johan', 'last': 'Jaeger'}))
test('2 < 1600', () => expect(2).toBeLessThan(1600))
test('no \'i\' in \'team\'', () => expect('team').not.toMatch(/i/)) /* regular-expression testing */
test('\'admin\' should be in array', () => expect(['Johan', 'Karen', 'admin']).toContain('admin')) 
test.skip('[a, b, c] has \'b\'', () => expect(['a', 'b', 'c']).toContain('b'))

test.todo('should not let numbers be passed.') /* a future test */ 

/* also test.only() and test.skip()... scope is local file. */

/* expect().toMatchSnapshot()... toMatchInlineSnapshot() is probably better... */

test('testing snapshots', () => {
    expect(1).toMatchSnapshot()
})

test('testing snapshots again', () => {
    expect(5).toMatchSnapshot()
})