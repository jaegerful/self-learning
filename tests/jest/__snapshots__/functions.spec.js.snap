// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing snapshots 1`] = `1`;

exports[`testing snapshots again 1`] = `5`;
